app/lib/application_logger.rb

```diff
@@ -0,0 +1,43 @@
+require "json"
+
+class ApplicationLogger < ActiveSupport::Logger
+  def debug(*msg, &block)
+    value = as_hash(msg[0], msg[1], &block)
+    super(value, &nil)
+  end
+
+  def info(*msg, &block)
+    value = as_hash(msg[0], msg[1], &block)
+    super(value, &nil)
+  end
+
+  def warn(*msg, &block)
+    value = as_hash(msg[0], msg[1], &block)
+    super(value, &nil)
+  end
+
+  def error(*msg, &block)
+    value = as_hash(msg[0], msg[1], &block)
+    super(value, &nil)
+  end
+
+  def fatal(*msg, &block)
+    value = as_hash(msg[0], msg[1], &block)
+    super(value, &nil)
+  end
+
+private
+
+  def as_hash(msg, attribs = {})
+    msg = yield if block_given?
+
+    message_to_hash(msg).merge(attribs || {}).merge(CurrentLoggingAttributes.attributes).compact
+  end
+
+  def message_to_hash(message)
+    return { message: } if message.is_a? String
+    return message if message.is_a? Hash
+
+    raise ArgumentError, "message must be a string or hash"
+  end
+end
```

Comment:
Feels like this could be a LogMessage class?

Reply:
Not sure I follow exactly what you're suggesting here. The intention here is that this class follows the same interface as the standard rails logger, where the first argument passed to `logger.info` etc. is a message, which can be of any type. This code deals with converting this to a hash and merging the logging context onto that hash before logging it - not sure which of this would benefit from extracting it out.